File: ihotel/ihotel/page/turndown/turndown.py
import frappe
from frappe.utils import today


@frappe.whitelist()
def get_turndown_data():
	"""
	Return guests who need turndown tonight:
	  1. Confirmed Reservations arriving today (primary focus)
	  2. Checked In guests with status Reserved or Checked In

	Turndown Status:
	  Completed    — a Housekeeping Task (Turndown, Completed) exists for the room today
	  Requested    — turndown_requested = 1 on the doc
	  Not Required — all others
	"""
	today_date = today()
	rows = []

	# ── 1. Confirmed Reservations arriving today ─────────────────────────────
	reservations = frappe.get_all(
		"Reservation",
		filters={"status": "confirmed", "check_in_date": today_date},
		fields=["name", "full_name", "room", "room_type", "turndown_requested", "turndown_reason"],
		order_by="room asc",
	)
	for res in reservations:
		room = (
			frappe.db.get_value("Room", res.room, ["room_number", "floor", "room_type"], as_dict=True)
			if res.room
			else None
		)
		rows.append({
			"source": "Reservation",
			"source_doc": res.name,
			"guest": res.full_name or res.name,
			"room": res.room or "",
			"room_number": room.room_number if room else (res.room or ""),
			"floor": room.floor if room else "",
			"room_type": room.room_type if room else (res.room_type or ""),
			"turndown_requested": int(res.turndown_requested or 0),
			"turndown_reason": res.turndown_reason or "",
		})

	# ── 2. Checked In guests (Reserved or Checked In status) ─────────────────
	stays = frappe.get_all(
		"Checked In",
		filters={"status": ["in", ["Reserved", "Checked In"]], "docstatus": 1},
		fields=["name", "guest", "room", "turndown_requested"],
		order_by="room asc",
	)
	for stay in stays:
		room = (
			frappe.db.get_value("Room", stay.room, ["room_number", "floor", "room_type"], as_dict=True)
			if stay.room
			else None
		)
		rows.append({
			"source": "Checked In",
			"source_doc": stay.name,
			"guest": stay.guest or "",
			"room": stay.room or "",
			"room_number": room.room_number if room else (stay.room or ""),
			"floor": room.floor if room else "",
			"room_type": room.room_type if room else "",
			"turndown_requested": int(stay.turndown_requested or 0),
			"turndown_reason": "",
		})

	# ── Completed turndown tasks for today ───────────────────────────────────
	completed_tasks = frappe.get_all(
		"Housekeeping Task",
		filters={
			"task_type": "Turndown",
			"status": "Completed",
			"assigned_date": [">=", today_date + " 00:00:00"],
		},
		fields=["room"],
	)
	completed_rooms = {t.room for t in completed_tasks}

	result = []
	for row in rows:
		if row["room"] in completed_rooms:
			row["turndown_status"] = "Completed"
		elif row["turndown_requested"]:
			row["turndown_status"] = "Requested"
		else:
			row["turndown_status"] = "Not Required"
		result.append(row)

	return result


@frappe.whitelist()
def toggle_turndown(doctype, docname, value):
	"""Toggle turndown_requested on a Reservation or Checked In document."""
	value = int(value)
	if doctype == "Reservation":
		frappe.db.set_value("Reservation", docname, "turndown_requested", value)
	elif doctype == "Checked In":
		frappe.db.set_value("Checked In", docname, "turndown_requested", value, update_modified=False)
	frappe.db.commit()
	return value

File: ihotel/ihotel/page/turndown/turndown.js
frappe.pages["turndown"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "Turndown",
		single_column: true,
	});

	page.main.html(`
		<div class="td-wrapper">
			<div class="td-pills-row">
				<button class="td-pill active" data-filter="">All <span class="td-pill-count">0</span></button>
				<button class="td-pill" data-filter="Requested"><span class="td-dot" style="background:#7c3aed;"></span>Requested <span class="td-pill-count">0</span></button>
				<button class="td-pill" data-filter="Completed"><span class="td-dot" style="background:#10b981;"></span>Completed <span class="td-pill-count">0</span></button>
				<button class="td-pill" data-filter="Not Required"><span class="td-dot" style="background:#9ca3af;"></span>Not Required <span class="td-pill-count">0</span></button>
			</div>
			<div class="td-container">
				<div class="td-loading">Loading…</div>
			</div>
		</div>
	`);

	const $container = page.main.find(".td-container");
	const $pills = page.main.find(".td-pill");
	let all_data = [];
	let active_filter = "";

	function apply_filter() {
		const rows = active_filter ? all_data.filter(r => r.turndown_status === active_filter) : all_data;
		render(rows);
	}

	function update_counts() {
		const counts = {};
		all_data.forEach(r => { counts[r.turndown_status] = (counts[r.turndown_status] || 0) + 1; });
		$pills.each((_, el) => {
			const f = $(el).data("filter");
			$(el).find(".td-pill-count").text(f ? (counts[f] || 0) : all_data.length);
		});
	}

	$pills.on("click", function () {
		$pills.removeClass("active");
		$(this).addClass("active");
		active_filter = $(this).data("filter");
		apply_filter();
	});

	function view_link(row) {
		if (row.source === "Reservation") {
			return `/app/reservation/${encodeURIComponent(row.source_doc)}`;
		}
		return `/app/checked-in/${encodeURIComponent(row.source_doc)}`;
	}

	function render(rows) {
		if (!rows.length) {
			$container.html('<div class="td-empty">No guests match this filter.</div>');
			return;
		}

		const status_colors = {
			"Requested":    "#7c3aed",
			"Completed":    "#10b981",
			"Not Required": "#9ca3af",
		};

		const html = `
			<table class="td-table">
				<thead>
					<tr>
						<th>Room</th>
						<th>Floor</th>
						<th>Room Type</th>
						<th>Guest</th>
						<th>Source</th>
						<th>Turndown Status</th>
						<th>Actions</th>
					</tr>
				</thead>
				<tbody>
					${rows.map(row => {
						const color = status_colors[row.turndown_status] || "#9ca3af";
						const source_badge = row.source === "Reservation"
							? `<span class="td-source-badge reservation">Reservation</span>`
							: `<span class="td-source-badge checked-in">Checked In</span>`;

						let action_btns = "";
						if (row.turndown_status !== "Completed") {
							if (row.turndown_requested) {
								action_btns += `<button class="td-toggle-btn btn btn-xs btn-warning" data-doctype="${frappe.utils.escape_html(row.source)}" data-docname="${frappe.utils.escape_html(row.source_doc)}" data-value="0">Cancel Turndown</button> `;
							} else {
								action_btns += `<button class="td-toggle-btn btn btn-xs btn-primary" data-doctype="${frappe.utils.escape_html(row.source)}" data-docname="${frappe.utils.escape_html(row.source_doc)}" data-value="1">Request Turndown</button> `;
							}
						}
						if (row.turndown_status === "Requested") {
							action_btns += `<button class="td-action-btn btn btn-xs btn-success" data-room="${frappe.utils.escape_html(row.room)}">Mark Complete</button> `;
						}
						action_btns += `<a href="${view_link(row)}" class="td-link">View</a>`;

						return `
							<tr>
								<td class="td-room">${frappe.utils.escape_html(row.room_number)}</td>
								<td>${frappe.utils.escape_html(row.floor || "—")}</td>
								<td>${frappe.utils.escape_html(row.room_type || "—")}</td>
								<td>${frappe.utils.escape_html(row.guest || "—")}</td>
								<td>${source_badge}</td>
								<td>
									<span class="td-status-badge" style="background:${color}20;color:${color};">
										${frappe.utils.escape_html(row.turndown_status)}
									</span>
									${row.turndown_reason ? `<div class="td-reason">${frappe.utils.escape_html(row.turndown_reason)}</div>` : ""}
								</td>
								<td>${action_btns}</td>
							</tr>
						`;
					}).join("")}
				</tbody>
			</table>
		`;
		$container.html(html);

		// Toggle turndown_requested
		$container.find(".td-toggle-btn").on("click", function () {
			const doctype  = $(this).data("doctype");
			const docname  = $(this).data("docname");
			const value    = $(this).data("value");
			const label    = value ? "requested" : "cancelled";
			frappe.call({
				method: "ihotel.ihotel.page.turndown.turndown.toggle_turndown",
				args: { doctype, docname, value },
				callback: () => {
					frappe.show_alert({ message: `Turndown ${label}`, indicator: value ? "purple" : "orange" });
					load();
				},
			});
		});

		// Mark Complete — creates a Housekeeping Task
		$container.find(".td-action-btn").on("click", function () {
			const room = $(this).data("room");
			frappe.call({
				method: "frappe.client.insert",
				args: {
					doc: {
						doctype: "Housekeeping Task",
						room: room,
						task_type: "Turndown",
						status: "Completed",
						assigned_date: frappe.datetime.now_datetime(),
						cleaned_date: frappe.datetime.now_datetime(),
					},
				},
				callback: () => {
					frappe.show_alert({ message: `Turndown for Room ${room} marked complete`, indicator: "green" });
					load();
				},
			});
		});
	}

	function load() {
		$container.html('<div class="td-loading">Loading…</div>');
		frappe.call({
			method: "ihotel.ihotel.page.turndown.turndown.get_turndown_data",
			callback: (r) => {
				all_data = r.message || [];
				update_counts();
				apply_filter();
			},
		});
	}

	page.set_secondary_action("Refresh", load);
	load();
};
